### PR TITLE
HLS output support (file mode)

### PR DESCRIPTION
**Pull request by anphsw**
This PR adds ability to write streams to HLS storage.
Will create playlists and *.ts files, which can be serverd by any web-server. Tested with nginx, apache, VLC and some STBs.
"hls=1" can be used with multicast=0 or unicast=0 as only writes files to disk (tmpfs preferred).

### CODE REVIEW

**Comment by Saentist** [comment]
Add options as:
Chunk size (MB) or chunk time (second's).
Playlist chunk count (number of chunks to keep in playlist/memory).

What about teletext pid problem?

**Comment by anphsw** [comment]
Just read Readme.
Chunk time = hls_rotate_time (default = 10).
Chunk count may be a problem here: what we should do, when not all chunks in the playlist ready? How we must form playlist right after start? Serve playlist with one chunk, then two, etc, or delay playlist generation for chunk_time * chunk count?
I think this is VOD option, it's not for the live stream.

Can you please describe teletext problem you mention?

**Comment by Saentist** [comment]
HLS playlist M3U8 solve count by regenerate playlist when new chunk added 
so when started each playlist can contains from one to maximum chunks

HLS is not a realtime streaming it's a delayed one, there is all time delay from live stream.

From where come Teletext problem or teletext sub's
if stream is 10 sec behind live and subs are delayed.
Players refuse to show someting from past.

For this case lot of comersial software extract teletext sub's as [WEBVTT](https://www.w3.org/TR/webvtt1/)
https://github.com/kanongil/telxcc 
https://github.com/CCExtractor/ccextractor
or similar software
 
Example in source of page ;)
https://developer.apple.com/videos/play/wwdc2017/504/

**Comment by anphsw** [comment]
Yes, dynamic chunks count is possible, but I cannot see real advantages of for example, 20,10,5 chunks vs 2. It's not priority task for me, but you can add such support by another PR.
Regarding subtitles - can we just rewrite time in subtitle pid to match actual stream? I don't find any channels with subtitles on my test box, will try to find them.

**Comment by Saentist** [comment]
> Yes, dynamic chunks count is possible, but I cannot see real advantages of for example, 20,10,5 chunks vs 2. It's not priority task for me, but you can add such support by another PR. Regarding subtitles - can we just rewrite time in subtitle pid to match actual stream? I don't find any channels with subtitles on my test box, will try to find them.

Chunk number will affect memory usage and I/O access, 
and on not so pontential devices, 
where IRQ is ignored by default will have problems.

More chunks can make channel start from hour after live stream.
with on other side is usefull for CATCHUP channels +1 +6 +12 etc
With some players can be used as timeshift. feature ;)

On Astra 19.2E have channels with teletext subtitles.

**Comment by anphsw** [comment]
Ok, I see you want timeshift. It's not in this PR anyway, but possible.

**Comment by Saentist** [comment]
> Ok, I see you want timeshift. It's not in this PR anyway, but possible.

No I dont want timeshift,
but if have 100 channels with average bitrate of 8 Mb/s
and have 4GB of memory, with must include TMPFS,
how many chunks are possuble before "no space left" message? 

In this case user will want to manage size and count of chunks.

This is my point, stable and unproblematic usage no matter hardware ;)

**Comment by anphsw** [comment]
So if you don't want timeshift, it's ok to have only two fully written chunks (as per default).
100 channels * 3 chunks * 8mbit / 8bits per byte = 300mb space per 1 second storage.
My test box have 60 channels (2-10mbit) from 10 mumudvb processes, 5 sec rotate time, and total consumed space is about 270mb. So I don't see problem here.

**Comment by anphsw** [comment]
To be clear, you cannot manage size of chunks if channels streamed in VBR, only possible solution is split by time.
Or else we must recalculate each chunk time based on its size and place to playlist. It's very hard with VBR.

**Comment by Saentist** [comment]
> To be clear, you cannot manage size of chunks if channels streamed in VBR, only possible solution is split by time. Or else we must recalculate each chunk time based on its size and place to playlist. It's very hard with VBR.

there is already done code for "segmenter" size cut
but beter one is IPB cuttig scenario. /cut before I frame [RFC8216](https://datatracker.ietf.org/doc/html/rfc8216)/
https://github.com/iliya-gr/mediasegmenter
https://github.com/xtne6f/tsmemseg
https://github.com/SteveCoul/MyBe

**Comment by anphsw** [comment]
I-frame cut is on the my roadmap, but it's not very useful, since I-frames from DVB is more frequent and chunk start without I-frame leads to only half second of damaged picture (same as in original unicast stream).
I think I will never implement size limitation, since you can clearly calculate maximum storage size based on input TP parameters. but you can write your own PR.

**Comment by Saentist** [comment]
Do not concentrate on imposible "for moment" things.
I cannot make PR, but can make suggestions/feature request as user and implementator.

As this one 
https://github.com/braice/MuMuDVB/issues/262

Now we wait for moment where this will be merged to MASTER

**Comment by anphsw** [comment]
Yes, I think now it's ready.

**Comment by trollcop** [comment]
@braice any particular thing blocking this PR?
Seems a fair bit of effort was spent on making it, and if it's not merged there's going to be some bitrot and functionality seems to be fairly useful? (Not too familiar with HLS tho).

I'm not sure why it makes changes to ts.h, but the rest of the stuff seems fairly well isolated from the rest of the code.
edit: never mind, ts.h changes are needed for hls.c

**Comment by anphsw** [comment]
Resolved merge conflicts; confirming 3+ month using DVB-S2/T2 input -> HLS output in production.

File: src/hls.c [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+		// match 2 (af + no payload) or 3 (af + payload) + RAI = 1
+		af=(af_header_t *)(buf + pos + TS_HEADER_LEN);
+    		if (header->adaptation_field_control > 1 && af->random_access_indicator)
+    		{
+		    log_message( log_module, MSG_FLOOD,"MPEG-TS RAI found at %d\n", pos);
+        	    goto exit;
+    		}
+
+		// h264 stream
+		// match 1 (payload only)
```
Comment (trollcop) [inline src/hls.c:76]: Is there a way to improve this to also detect MPEG2/h265 i-frame start codes?

Comment (anphsw) [reply src/hls.c:76]: Can't find understandable documentation on which types of h265 i-frames cutting is allowed.
And in theory, MPEG2 encode must convert i-frames from high-level protocol to RAI by itself. In practice, there is rare case when encoder does not support RAI, but support i-frame.

**Comment by braice** [comment]
More an overlook than anything else, I will look at it

File: src/ts.h [hunk 2]
```diff
@@ -190,6 +190,32 @@ typedef struct {
 #endif
 } ts_header_t;
 
+typedef struct {
+  u_char adaptation_field_length		:8;
+#if BYTE_ORDER == BIG_ENDIAN
```
Comment (trollcop) [inline src/ts.h]: This should be changed to `#ifdef __BIG_ENDIAN__` to match the rest of the file - and actually, the correct way to determine endianness (in gcc at least) is to check `__BYTE_ORDER__` define, so both are wrong.  Actually, I changed the `#if BYTE_ORDER == BIG_ENDIAN` (which was also incorrect, and depends on non-portable `endian.h` in my commit a few weeks back, so I think the least amount of changes path would be to:

```
#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
#define __BIG_ENDIAN__
#endif
```

at top of ts.h and keep the rest of the file unchanged, and fix this line 195 to same `#ifdef __BIG_ENDIAN__`.
And then of course some testing on BE system, I'm guessing this is a leftover of when someone tried to build this on OpenWRT or similar.

File: src/hls.c (hunk 1 above)
Comment (trollcop) [reply src/hls.c:76]: Ah, I see. Makes sense, so normally encoder would just set random access indicator, and the code breaks out at that point. And if that's not available, then manually search for i-frame start based on encoding type.

File: src/ts.h (hunk 2 above)
Comment (anphsw) [reply src/ts.h]: Changed file to match current syntax.
But never tested on big endian due to lack of hardware.

**Comment by braice** [comment]
Let's go for it !

**Comment by Saentist** [comment]
@anphsw 
Is it posible to add simple example in examples folder?

**Pull request merged**